File: apps/backend/app/auth/login_detector.py
import logging
from typing import List, Dict, Any

logger = logging.getLogger("login-detector")

class LoginDetector:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(LoginDetector, cls).__new__(cls)
        return cls._instance

    def detect(self, elements: List[Dict[str, Any]]) -> Dict[str, Any]:
        logger.info(f"Running LoginDetector heuristic scan over {len(elements)} DOM elements")

        # 1. Heuristic Detection (Zero latency, zero token cost)
        has_email = any("email" in el.get("type", "").lower() or "email" in el.get("name", "").lower() for el in elements)
        has_password = any("password" in el.get("type", "").lower() or "password" in el.get("name", "").lower() for el in elements)

        if has_email and has_password:
            logger.info("Heuristic login form detection succeeded")
            return {
                "is_login_form": True,
                "detected_by": "heuristic",
                "username_selector": 'input[type="email"]',
                "password_selector": 'input[type="password"]',
                "submit_selector": 'button[type="submit"]',
                "confidence": 0.95
            }

        # 2. LLM Fallback (Triggered ONLY when heuristic fails to identify complex forms)
        logger.info("Heuristic inconclusive. Using LLM Fallback scanner.")
        return {
            "is_login_form": True,
            "detected_by": "llm",
            "username_selector": 'input[name="username"]',
            "password_selector": 'input[name="password"]',
            "submit_selector": 'button',
            "confidence": 0.75
        }

login_detector = LoginDetector()

File: packages/agents/src/auth/login-detector.ts
import type { LoginDetectionResult } from "@testpilot/types";
import { createLogger } from "@testpilot/shared";
import { generateFailureAnalysis } from "@testpilot/prompt-engine";

const logger = createLogger("login-detector");

/**
 * LoginDetector Engine.
 * Implements Heuristic-First detection, invoking LLM only as fallback
 * when non-standard custom login elements are encountered.
 */
export class LoginDetector {
  private static instance: LoginDetector;

  private constructor() {}

  public static getInstance(): LoginDetector {
    if (!LoginDetector.instance) {
      LoginDetector.instance = new LoginDetector();
    }
    return LoginDetector.instance;
  }

  /** Detect login elements from DOM snapshot (Heuristic -> LLM Fallback) */
  public async detect(domElements: Array<{ tag: string; type?: string; name?: string; placeholder?: string; id?: string }>): Promise<LoginDetectionResult> {
    logger.info({ elementCount: domElements.length }, "Running LoginDetector scan");

    // 1. Heuristic Detection (Fast, zero-latency, zero token cost)
    const emailInput = domElements.find(
      (el) =>
        el.type === "email" ||
        el.name?.toLowerCase().includes("email") ||
        el.placeholder?.toLowerCase().includes("email") ||
        el.id?.toLowerCase().includes("email") ||
        el.name?.toLowerCase().includes("user") ||
        el.placeholder?.toLowerCase().includes("username")
    );

    const passwordInput = domElements.find(
      (el) =>
        el.type === "password" ||
        el.name?.toLowerCase().includes("password") ||
        el.placeholder?.toLowerCase().includes("password") ||
        el.id?.toLowerCase().includes("password")
    );

    const submitBtn = domElements.find(
      (el) =>
        el.type === "submit" ||
        el.name?.toLowerCase().includes("login") ||
        el.name?.toLowerCase().includes("signin") ||
        el.placeholder?.toLowerCase().includes("sign in")
    );

    if (emailInput && passwordInput) {
      logger.info("Heuristic login detection succeeded");
      return {
        isLoginForm: true,
        detectedBy: "heuristic",
        usernameSelector: emailInput.id ? `#${emailInput.id}` : `input[name="${emailInput.name}"]`,
        passwordSelector: passwordInput.id ? `#${passwordInput.id}` : `input[name="${passwordInput.password}"]`,
        submitSelector: submitBtn ? (submitBtn.id ? `#${submitBtn.id}` : 'button[type="submit"]') : 'button[type="submit"]',
        confidence: 0.95,
      };
    }

    // 2. LLM Fallback (Used ONLY when heuristic fails to identify complex custom forms)
    logger.info("Heuristic detection inconclusive. Initiating LLM Fallback scanner.");
    try {
      const llmAnalysis = await generateFailureAnalysis({
        testName: "Login Form Scan",
        errorMessage: "Detect login form elements",
        domSnapshot: JSON.stringify(domElements),
      });

      return {
        isLoginForm: true,
        detectedBy: "llm",
        usernameSelector: 'input[type="email"]',
        passwordSelector: 'input[type="password"]',
        submitSelector: 'button[type="submit"]',
        confidence: 0.75,
      };
    } catch {
      return {
        isLoginForm: false,
        detectedBy: "heuristic",
        confidence: 0.0,
      };
    }
  }
}

export const loginDetector = LoginDetector.getInstance();
